src/aggregate/reducers/to_list.c

```diff
@@ -6,75 +6,70 @@
 
 #include <aggregate/reducer.h>
 
-typedef struct {
-  TrieMap *values;
-  const RLookupKey *srckey;
-} tolistCtx;
+static uint64_t hashFunction_RSValue(const void *key) {
+  return RSValue_Hash(key, 0);
+}
+static void *dup_RSValue(void *p, const void *key) {
+  return RSValue_IncrRef((RSValue *)key);
+}
+static int compare_RSValue(void *privdata, const void *key1, const void *key2) {
+  return RSValue_Equal(key1, key2, NULL);
+}
+static void destructor_RSValue(void *privdata, void *val) {
+  RSValue *v = val;
+  RSValue_Decref(v);
+}
+
+static dictType RSValueSet = {
+  .hashFunction = hashFunction_RSValue,
+  .keyDup = dup_RSValue,
+  .valDup = NULL,
+  .keyCompare = compare_RSValue,
+  .keyDestructor = destructor_RSValue,
+  .valDestructor = NULL,
+};
 
 static void *tolistNewInstance(Reducer *rbase) {
-  tolistCtx *ctx = Reducer_BlkAlloc(rbase, sizeof(*ctx), 100 * sizeof(*ctx));
-  ctx->values = NewTrieMap();
-  ctx->srckey = rbase->srckey;
-  return ctx;
+  dict *values = dictCreate(&RSValueSet, NULL);
+  return values;
 }
 
 static int tolistAdd(Reducer *rbase, void *ctx, const RLookupRow *srcrow) {
-  tolistCtx *tlc = ctx;
-  RSValue *v = RLookup_GetItem(tlc->srckey, srcrow);
+  dict *values = ctx;
+  RSValue *v = RLookup_GetItem(rbase->srckey, srcrow);
   if (!v) {
     return 1;
   }
 
   // for non array values we simply add the value to the list */
   if (v->t != RSValue_Array) {
-    uint64_t hval = RSValue_Hash(v, 0);
-    if (TrieMap_Find(tlc->values, (char *)&hval, sizeof(hval)) == TRIEMAP_NOTFOUND) {
-
-      TrieMap_Add(tlc->values, (char *)&hval, sizeof(hval),
-                  RSValue_IncrRef(RSValue_MakePersistent(v)), NULL);
-    }
+    dictAdd(values, v, NULL);
   } else {  // For array values we add each distinct element to the list
     uint32_t len = RSValue_ArrayLen(v);
     for (uint32_t i = 0; i < len; i++) {
-      RSValue *av = RSValue_ArrayItem(v, i);
-      uint64_t hval = RSValue_Hash(av, 0);
-      if (TrieMap_Find(tlc->values, (char *)&hval, sizeof(hval)) == TRIEMAP_NOTFOUND) {
-
-        TrieMap_Add(tlc->values, (char *)&hval, sizeof(hval),
-                    RSValue_IncrRef(RSValue_MakePersistent(av)), NULL);
-      }
+      dictAdd(values, RSValue_ArrayItem(v, i), NULL);
     }
   }
   return 1;
 }
 
 static RSValue *tolistFinalize(Reducer *rbase, void *ctx) {
-  tolistCtx *tlc = ctx;
-  TrieMapIterator *it = TrieMap_Iterate(tlc->values, "", 0);
-  char *c;
-  tm_len_t l;
-  RSValue *val;
-  RSValue **arr = RSValue_AllocateArray(tlc->values->cardinality);
-  size_t i = 0;
-  while (TrieMapIterator_Next(it, &c, &l, (void **)&val)) {
-    if (val) {
-      arr[i++] = RSValue_IncrRef(val);
-    }
+  dict *values = ctx;
+  size_t len = dictSize(values);
+  dictIterator *it = dictGetIterator(values);
```

Comment:
anything can be affected here? list creation order?

Reply:
Previously the order was lexicographic over the hash value of the values when viewed as a char[8]. I think the dictionary also iterates over the values according to the hash value, but there was no guarantee for any specific order anyway